File: 4 kyu/Sum of Intervals/solution.ex
defmodule Kata do
  def sum_of_intervals(intervals) do
    intervals
      |> Enum.sort
      |> Stream.scan(fn {b, e}, {_, x} -> {max(b, x), max(e, x)} end)
      |> Stream.map(fn {b, e} -> e - b end)
      |> Enum.sum
  end
end
_______________________
defmodule Kata do
  def sum_of_intervals(a) do
    a
    |> Enum.map(&do_get_intervals(&1))
    |> List.flatten()
    |> Enum.uniq()
    |> length()
  end
  
  defp do_get_intervals({first, last}) do
    Enum.map(first..last-1, &(&1))
  end
end
______________________
defmodule Kata do
  def sum_of_intervals(xs), do: MapSet.size(walk(MapSet.new(), xs))
  defp walk(res, []), do: res
  defp walk(res, [x | xs]), do: walk(walk2(res, elem(x, 0), elem(x, 1)), xs)
  defp walk2(res, i, b) when i == b, do: res
  defp walk2(res, i, b), do: walk2(MapSet.put(res, i), i + 1, b)
end
___________________________
defmodule Kata do
  def sum_of_intervals(a) do
    Enum.reduce(a, MapSet.new, &add_range/2)
    |> MapSet.size
  end
  
  def add_range({x, y}, acc) when x == y, do: acc
  def add_range({x, y}, acc), do: add_range({x+1, y}, MapSet.put(acc, {x, x+1}))
end
_________________
defmodule Kata, do: def sum_of_intervals(a), do: a |> Enum.flat_map(fn x -> elem(x, 0)..elem(x, 1) - 1 end) |> Enum.uniq |> Enum.count

File: 4 kyu/Sum of Intervals/solution.clj
(ns sum-intervals.core)

(defn sum-intervals
  [intervals]
  (->> intervals
       (mapcat #(apply range %))
       set
       count))

________________
(ns sum-intervals.core)

(defn overlap? [[x y] [i j]] (not (or (> i y) (< j x))))
   
(defn merge-interval
  "Merge interval to the end of a sorted vector of intervals"
  [intervals [x y :as interval]]
  (let [[i j :as last-int] (last intervals)]
    (if (overlap? interval last-int)
      (conj (vec (butlast intervals)) [(min x i) (max y j)])
      (conj intervals interval)
    )
  )
)

(defn interval-length [[a b]] (- b a))

(defn sum-intervals
  [intervals]
  (->> intervals
       (sort)
       (reduce merge-interval [[0 0]])
       (map interval-length)
       (reduce +)
  )
)

_________________________
(ns sum-intervals.core)

(defn sum-intervals [intervals]
  (->> intervals
    sort
    (reductions (fn [[_b' e'] [b e]] [(max b e') (max e e')]))
    (map (fn [[b e]] (- e b)))
    (reduce +)))

___________________
(ns sum-intervals.core)

(defn sum-intervals [intervals]
  (count (set (flatten (for [interval intervals]
    (for [i (range (first interval) (second interval))]
      i))))))

______________________
(ns sum-intervals.core)

;; Maps intervals onto a number line, tracking changes in the number of open intervals
(defn to-numline [intervals]
  (-> (fn [numline intv] 
        (-> numline 
          (assoc (first intv) (+ 1 (numline (first intv) 0))) 
          (assoc (last intv) (- (numline (last intv) 0) 1)))) 
    (reduce {} intervals)))

;; Loops over numline keys, summing when all intervals have closed
(defn sum-intervals [intervals]
  (let [numline (to-numline intervals)]
    (loop [ks (sort (keys numline))
           opener nil
           score 0
           sum 0]
      (let [k (first ks)
            d (numline k)]
        (cond
          (nil? k) sum
          (= 0 score) (recur (rest ks) k (+ score d) sum)
          (= 0 (+ score d)) (recur (rest ks) nil 0 (+ sum (- k opener)))
          :else (recur (rest ks) opener (+ score d) sum))))))

____________________________
(ns sum-intervals.core)

(defn sum-intervals [intervals]
  (->> intervals
       (sort)
       (reductions (fn  [[a b] [c d]] [(max b c) (max b d)]))
       (map (fn [[x y]] (- y x)))
       (reduce +)))

_______________
(ns sum-intervals.core)

(defn r
  [[a b]]
  (range a b))

(defn sum-intervals
  [intervals]
  (count (reduce #(apply conj %1 (r %2)) #{} intervals)))

_____________________
(ns sum-intervals.core)

(defn sum-intervals
  [intervals]
  (loop [sl intervals
         all-ranges []]
    (if (zero? (count sl)) (count (set (flatten all-ranges)))
        (recur (next sl) (conj all-ranges (range (first (first sl)), (second (first sl))))))))
